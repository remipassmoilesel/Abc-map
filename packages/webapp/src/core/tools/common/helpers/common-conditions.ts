/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import MapBrowserEvent from 'ol/MapBrowserEvent';
import { FeatureLike } from 'ol/Feature';
import { SupportedGeometry } from '../interactions/SupportedGeometry';

export function withShiftKey(ev: MapBrowserEvent<UIEvent | KeyboardEvent | MouseEvent | TouchEvent>): boolean {
  const original = ev.originalEvent;
  return 'shiftKey' in original && original.shiftKey;
}

export function withControlKey(ev: MapBrowserEvent<UIEvent | KeyboardEvent | MouseEvent | TouchEvent>): boolean {
  const original = ev.originalEvent;
  return 'ctrlKey' in original && original.ctrlKey;
}

export function withControlKeyOnly(ev: MapBrowserEvent<UIEvent | KeyboardEvent | MouseEvent | TouchEvent>): boolean {
  return withControlKey(ev) && !withShiftKey(ev) && !withAltKey(ev);
}

export function withAltKey(ev: MapBrowserEvent<UIEvent | KeyboardEvent | MouseEvent | TouchEvent>): boolean {
  const original = ev.originalEvent;
  return 'altKey' in original && original.altKey;
}

export function withGeometryOfType(feat: FeatureLike, types: SupportedGeometry[]): boolean {
  const type = feat?.getGeometry()?.getType();
  return (type && !!types.find((t) => t === type)) || false;
}

export function withGeometry(feat: FeatureLike): boolean {
  return !!feat?.getGeometry()?.getType();
}
