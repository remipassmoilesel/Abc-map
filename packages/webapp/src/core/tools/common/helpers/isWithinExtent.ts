/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { containsXY, Extent } from 'ol/extent';

export function isWithinExtent(selection: Extent, geomExtent?: Extent) {
  return geomExtent && containsXY(selection, geomExtent[0], geomExtent[1]) && containsXY(selection, geomExtent[2], geomExtent[3]);
}
