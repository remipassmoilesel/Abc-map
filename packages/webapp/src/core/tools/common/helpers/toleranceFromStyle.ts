/**
 * Copyright © 2021 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { Geometry } from 'ol/geom';
import Feature from 'ol/Feature';
import { FeatureWrapper } from '../../../geo/features/FeatureWrapper';

export function toleranceFromStyle(feature: Feature<Geometry>, resolution: number): number {
  const styleProps = FeatureWrapper.from(feature).getStyleProperties();
  const pointSize = (styleProps.point?.size || 0) / 2;
  const strokeWidth = (styleProps.stroke?.width || 0) / 2;
  return (pointSize + strokeWidth) * resolution;
}
