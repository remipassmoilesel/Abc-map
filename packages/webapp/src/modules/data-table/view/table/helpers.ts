/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { DataRow } from '../../../../core/data/data-source/DataSource';
import { asNumberOrString } from '../../../../core/utils/numbers';
import { getRemSize } from '../../../../core/ui/getRemSize';
import { getTextWidth } from '../../../../core/ui/getTextWidth';

// Computes column sizes depending on larger values
export function getAllColumSizes(rows: DataRow[]): { [k: string]: number } {
  // k: fieldName, v: maximum length as string
  const result: { [k: string]: number } = {};
  for (const row of rows) {
    for (const field in row.data) {
      let value = asNumberOrString(row.data[field]) + '';
      // If value is shorter than field name, we use field name as base
      if (!value || value.length < field.length) {
        value = field;
      }

      // We keep only the higher value
      result[field] = Math.max(getColumnSize(value), result[field] || 0);
    }
  }

  return result;
}

export function getColumnSize(value: string | number) {
  const remSize = getRemSize();
  const filterIconSize = 2 * remSize;
  const margins = 3.5 * remSize;

  return Math.round(Math.max(getTextWidth(value + '') + filterIconSize + margins));
}
