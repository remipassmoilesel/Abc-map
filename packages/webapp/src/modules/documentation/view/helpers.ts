/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 *
 *
 *
 */

const indexRegex = /index.html/gi;

/**
 * This function transform a webapp documentation URL to a static documentation URL
 *
 * @param pathname
 */
export function rewriteContentPath(pathname: string): string {
  const parts = pathname.split('/').filter((v) => !!v.trim() && !v.match(indexRegex));
  const lang = parts[0];
  return withTrailingSlash('/documentation/' + lang + '/' + parts.slice(3, parts.length).join('/'));
}

/**
 * This function recursively search assets in "content" and rewrite their source URL to match
 * webapp route scheme
 *
 * @param content
 * @param contentRoute
 */
export function rewriteAssetsUrls(content: Element, contentRoute: string) {
  const elements = content.querySelectorAll('img,video') as NodeListOf<HTMLVideoElement | HTMLImageElement>;
  for (const element of elements) {
    if (element.tagName === 'IMG' || element.tagName === 'VIDEO') {
      const existingSrc = element.getAttribute('src');
      element.src = contentRoute + existingSrc;
    }
  }
}

export function isExternalURL(url: string) {
  const _url = url.trim().toLowerCase();
  return _url.startsWith('http') || _url.startsWith('www');
}

function withTrailingSlash(str: string): string {
  if (!str.trim().endsWith('/')) {
    return str.trim() + '/';
  }

  return str;
}
