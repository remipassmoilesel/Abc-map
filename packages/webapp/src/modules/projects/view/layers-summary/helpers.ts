/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { LayerType } from '@abc-map/shared';
import { prefixedTranslation } from '../../../../i18n/i18n';
import { IconDefs } from '../../../../components/icon/IconDefs';
import { IconDefinition } from '@fortawesome/fontawesome-svg-core';

const t = prefixedTranslation('ProjectManagement:');

export function layerTypeName(type: LayerType | undefined): string {
  switch (type) {
    case LayerType.Vector:
      return t('Vector');
    case LayerType.Predefined:
      return t('Predefined');
    case LayerType.Wms:
      return t('Wms');
    case LayerType.Wmts:
      return t('Wmts');
    case LayerType.Xyz:
      return t('Xyz');
    default:
      return t('Unknown_type');
  }
}

export function layerIcon(type: LayerType | undefined): IconDefinition {
  switch (type) {
    case LayerType.Vector:
      return IconDefs.faDrawPolygon;
    case LayerType.Predefined:
      return IconDefs.faClone;
    case LayerType.Wms:
      return IconDefs.faClone;
    case LayerType.Wmts:
      return IconDefs.faClone;
    case LayerType.Xyz:
      return IconDefs.faClone;
    default:
      return IconDefs.faQuestion;
  }
}
