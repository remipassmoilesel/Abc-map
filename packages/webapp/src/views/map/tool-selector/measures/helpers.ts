/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { getLength, getArea } from 'ol/sphere';
import { AbcProjection } from '@abc-map/shared';
import Geometry from 'ol/geom/Geometry';
import { getNumberFormatter } from '../../../../i18n/i18n';

export function formatLength(geometry: Geometry, projection?: AbcProjection): string {
  const length = getLength(geometry, { projection: projection?.name });
  const formatter = getNumberFormatter();

  let output;
  if (length > 100) {
    output = `${formatter.format(Math.round((length / 1000) * 100) / 100)} km`;
  } else {
    output = `${formatter.format(Math.round(length * 100) / 100)} m`;
  }
  return output;
}

export function formatArea(geometry: Geometry, projection?: AbcProjection): string {
  const area = getArea(geometry, { projection: projection?.name });
  const formatter = getNumberFormatter();

  let output;
  if (area > 10000) {
    output = `${formatter.format(Math.round((area / 1000000) * 100) / 100)} km²`;
  } else {
    output = `${formatter.format(Math.round(area * 100) / 100)} m²`;
  }
  return output;
}
