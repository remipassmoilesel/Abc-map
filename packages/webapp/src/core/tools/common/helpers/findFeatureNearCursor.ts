/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import MapBrowserEvent from 'ol/MapBrowserEvent';
import { Geometry } from 'ol/geom';
import Feature from 'ol/Feature';
import VectorSource from 'ol/source/Vector';
import { toleranceFromStyle } from './toleranceFromStyle';
import { isCloseTo } from './isCloseTo';
import { DefaultTolerancePx } from '../constants';

export declare type FeatureFilter = (f: Feature<Geometry>) => boolean;
export const noopFilter: FeatureFilter = () => true;

/**
 * Find a feature under cursor if any, or near cursor.
 *
 * We must seek features around cursor for 'thin' geometries like points or lines.
 *
 * /!\ Vector source parameter must have useSpatialIndex option enabled
 *
 */
export function findFeatureNearCursor(
  event: MapBrowserEvent<MouseEvent>,
  source: VectorSource<Geometry>,
  filter = noopFilter,
  tolerancePx = DefaultTolerancePx
): Feature<Geometry> | undefined {
  const coord = event.coordinate;

  // First we try to find a feature on coordinates
  const features = source.getFeaturesAtCoordinate(coord).filter(filter);
  if (features.length) {
    return features[0];
  }

  // If nothing found, we try to find a feature close to coordinates
  // This code allow user to select small points
  const feature = source.getClosestFeatureToCoordinate(coord) as Feature<Geometry> | undefined; // getClosestFeatureToCoordinate typing is borked
  const closestPoint = feature?.getGeometry()?.getClosestPoint(coord);
  if (!feature || !closestPoint) {
    return;
  }

  const resolution = event.map.getView().getResolution() || 1;
  const tolerance = toleranceFromStyle(feature, resolution) + tolerancePx * resolution;

  const matchFilter = feature && filter(feature);
  const closeTo = isCloseTo(closestPoint, coord, tolerance);
  return (matchFilter && closeTo && feature) || undefined;
}
