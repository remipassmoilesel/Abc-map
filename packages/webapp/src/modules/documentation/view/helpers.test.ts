/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 *
 *
 *
 */

import { rewriteAssetsUrls, rewriteContentPath } from './helpers';

describe('helpers', () => {
  it('rewriteContentPath()', () => {
    expect(rewriteContentPath('/fr/modules/documentation')).toEqual('/documentation/fr/');
    expect(rewriteContentPath('/en/modules/documentation')).toEqual('/documentation/en/');
    expect(rewriteContentPath('/fr/modules/documentation/')).toEqual('/documentation/fr/');
    expect(rewriteContentPath('/fr/modules/documentation/reference/presentation')).toEqual('/documentation/fr/reference/presentation/');
    expect(rewriteContentPath('/fr/modules/documentation/reference/presentation/index.html')).toEqual('/documentation/fr/reference/presentation/');
  });

  it('rewriteAssetsUrls()', () => {
    // Prepare
    const content = document.createElement('div');

    const img = document.createElement('img');
    img.src = 'image.png';
    content.append(img);

    const video = document.createElement('video');
    video.src = 'video.mp4';
    content.append(video);

    const contentRoute = '/documentation/en/automated-test-target/';

    // Act
    rewriteAssetsUrls(content, contentRoute);

    // Assert
    expect(content.innerHTML).toMatchSnapshot();
  });
});
