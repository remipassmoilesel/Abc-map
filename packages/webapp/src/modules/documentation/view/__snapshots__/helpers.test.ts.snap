// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers rewriteAssetsUrls() 1`] = `"<img src=\\"/documentation/en/automated-test-target/image.png\\"><video src=\\"/documentation/en/automated-test-target/video.mp4\\"></video>"`;
